cmd/gateway-gcs.go

```diff
@@ -561,266 +580,169 @@ func (l *gcsGateway) DeleteObject(bucket string, object string) error {
-			break
-		}
+	object = l.client.Bucket(bucket).Object(dataName)
 
-		attrs, err := it.Next()
-		if err == iterator.Done {
-			break
-		} else if err != nil {
-			return ListPartsInfo{}, gcsToObjectError(traceError(err), bucket, prefix)
+	w := object.NewWriter(l.ctx)
+	if md5Hex != "" {
+		var err error
+		w.MD5, err = hex.DecodeString(md5Hex)
+		if err != nil {
+			return PartInfo{}, gcsToObjectError(traceError(err), bucket, key)
 		}
+	}
+	_, err = io.Copy(w, teeReader)
+	if err != nil {
+		return PartInfo{}, gcsToObjectError(traceError(err), bucket, key)
+	}
 
-		if attrs.Prefix != "" {
-			continue
-		}
+	err = w.Close()
+	if err != nil {
+		fmt.Println(err)
+		return PartInfo{}, gcsToObjectError(traceError(err), bucket, key)
+	}
 
-		_, _, partID, err := fromGCSMultipartKey(attrs.Name)
-		if err != nil {
-			continue
-		} else if partID == 0 {
-			// we'll ignore partID 0, it is our zero object, containing
-			// metadata
-			continue
+	if sha256sum != "" {
+		if hex.EncodeToString(sha256Writer.Sum(nil)) != sha256sum {
+			object.Delete(l.ctx)
+			return PartInfo{}, traceError(SHA256Mismatch{})
 		}
+	}
 
-		nextPartnumberMarker = partID
+	return PartInfo{
+		PartNumber:   partID,
+		ETag:         etag,
+		LastModified: time.Now().UTC(),
+		Size:         size,
+	}, err
+}
 
-		parts = append(parts, PartInfo{
-			PartNumber:   partID,
-			LastModified: attrs.Updated,
-			ETag:         fmt.Sprintf("%d", attrs.CRC32C),
-			Size:         attrs.Size,
-		})
+// ListObjectParts returns all object parts for specified object in specified bucket
+func (l *gcsGateway) ListObjectParts(bucket string, key string, uploadID string, partNumberMarker int, maxParts int) (ListPartsInfo, error) {
+	meta := gcsMultipartMetaName(uploadID)
+	object := l.client.Bucket(bucket).Object(meta)
+
+	_, err := object.Attrs(l.ctx)
+	if err != nil {
+		return ListPartsInfo{}, gcsToObjectError(traceError(err), bucket, key)
 	}
 
-	return ListPartsInfo{
-		IsTruncated:          isTruncated,
-		NextPartNumberMarker: nextPartnumberMarker,
-		Parts:                parts,
-	}, nil
+	return ListPartsInfo{}, nil
 }
 
-// AbortMultipartUpload aborts a ongoing multipart upload
-func (l *gcsGateway) AbortMultipartUpload(bucket string, key string, uploadID string) error {
-	delimiter := slashSeparator
-	prefix := pathJoin(ZZZZMinioPrefix, fmt.Sprintf("multipart-%s-%s", escape(key), uploadID))
+// Called by AbortMultipartUpload and CompleteMultipartUpload for cleaning up.
+func (l *gcsGateway) cleanupMultipartUpload(bucket, key, uploadID string) error {
+	meta := gcsMultipartMetaName(uploadID)
+	object := l.client.Bucket(bucket).Object(meta)
 
-	// iterate through all parts and delete them
-	it := l.client.Bucket(bucket).Objects(l.ctx, &storage.Query{Delimiter: delimiter, Prefix: prefix, Versions: false})
+	_, err := object.Attrs(l.ctx)
+	if err != nil {
+		return gcsToObjectError(traceError(err), bucket, key)
+	}
+
+	prefix := fmt.Sprintf("%s/multipart/%s/", gcsMinioMeta, uploadID)
```

Comment:
This multipart can become a string constant i guess.. since its static either can be part of existing constants like gcsMinioMeta, gcsMinioMetaMultipart?